### PR TITLE
Errors management

### PR DESCRIPTION
**Pull request by ImCRV**
A big refactoring pass introducing a better error management.
All Vulkan errors are raised as exceptions, simplifying this first step (this way, we ensure it will be "returned" to the caller, even if we do not manage errors properly in between).
Still work to do in future refactoring passes.

### CODE REVIEW

File: src/vulkan/ImVkCore.hpp [hunk 1]
```diff
@@ -106,74 +112,41 @@ struct Vk_OffscreenPass
...
     if (sampler)
     {
-      IM_ASSERT(m_vk);
-      m_vk->vkDestroySampler(owner_device, sampler, 0);
-      sampler = 0;
+      m_vk->vkDestroySampler(owner_device, sampler, nullptr);
+      sampler = VK_NULL_HANDLE;
     }
 
-    if (color.image && color.m_isImageOwner)
```
Comment (robin-imverse) [inline src/vulkan/ImVkCore.hpp:132]: Where did the m_isImageOwner test go? We need that as with the use of interoperability I added the possibility of assigning (and not reallocating) a color/depth image. If it is assigned we shouldn't destroy it here.

File: src/vulkan/ImVkCore.hpp [hunk 2]
```diff
@@ -106,74 +112,41 @@ struct Vk_OffscreenPass
...
-    {
-      IM_ASSERT(m_vk);
-      m_vk->vkDestroyImage(owner_device, color.image, 0);
-      m_vk->vkFreeMemory(owner_device, color.mem, 0);
-    }
+    color.clean(*m_vk, owner_device);
+    depth.clean(*m_vk, owner_device);
 
-    if (color.view && color.m_isImageViewOwner)
```
Comment (robin-imverse) [inline src/vulkan/ImVkCore.hpp:139]: same comment as above

File: src/vulkan/ImVkCore.hpp [hunk 3]
```diff
@@ -106,74 +112,41 @@ struct Vk_OffscreenPass
...
+      ta.clean(*m_vk, owner_device);
     }
-
-    color.image = 0;
-    color.mem = 0;
-    color.view = 0;
-    color.m_isImageOwner = 0;
-    color.m_isImageViewOwner = 0;
-
-    if (depth.image && depth.m_isImageOwner)
```
Comment (robin-imverse) [inline src/vulkan/ImVkCore.hpp:150]: same comment as above

File: src/vulkan/ImVkCore.hpp [hunk 4]
```diff
@@ -106,74 +112,41 @@ struct Vk_OffscreenPass
...
-    {
-      IM_ASSERT(m_vk);
-      m_vk->vkDestroyImage(owner_device, depth.image, 0);
-      m_vk->vkFreeMemory(owner_device, depth.mem, 0);
-    }
-
-    if (depth.view && depth.m_isImageViewOwner)
```
Comment (robin-imverse) [inline src/vulkan/ImVkCore.hpp:157]: same comment as above

File: src/ImVkTexture.cpp [hunk 5]
```diff
@@ -756,6 +583,7 @@ bool ImVkTexture::updateTextureOnGPU(
 
   updating_mutex.lock();
 
+  // TODO: WTF?!
```
Comment (robin-imverse) [inline src/ImVkTexture.cpp:586]: Those states are actually old. They were previously used to know when projection mapping textures were ready to be release ore reused. We could remove completely current_use_state I believe.

File: src/ImVkMemorySharedImage.cpp [hunk 6]
```diff
@@ -572,30 +579,28 @@ VkResult ImVkMemorySharedImage::vkAllocateMemoryExternalGL(
...
-  if (result_mapping != VK_SUCCESS)
+  result = m_vk->vkGetMemoryWin32HandleKHR(device, &handle_infos, &handle);
+  if (result != VK_SUCCESS)
   {
-    ImLog("vkGetMemoryWin32HandleKHR_EXT failed with error = %d", result_mapping);
+    ImLog("vkGetMemoryWin32HandleKHR_EXT failed with error = %d", result);
+    return result;
   }
   
 #ifdef USE_GL_INTEROP
```
Comment (robin-imverse) [inline src/ImVkMemorySharedImage.cpp:589]: Should that define go away completely?

File: src/vulkan/ImVkCore.hpp (hunk 1 above)
Comment (ImCRV) [reply src/vulkan/ImVkCore.hpp:132]: Everything is managed in Vk_TextureAttachment structure now (ImVkCore.hpp)

File: src/vulkan/ImVkCore.hpp (hunk 2 above)
Comment (ImCRV) [reply src/vulkan/ImVkCore.hpp:139]: same answer as above

File: src/vulkan/ImVkCore.hpp (hunk 3 above)
Comment (ImCRV) [reply src/vulkan/ImVkCore.hpp:150]: same answer as above

File: src/vulkan/ImVkCore.hpp (hunk 4 above)
Comment (ImCRV) [reply src/vulkan/ImVkCore.hpp:157]: same answer as above

File: src/ImVkTexture.cpp (hunk 5 above)
Comment (ImCRV) [reply src/ImVkTexture.cpp:586]: I note this modification, I will do in my next refactoring pass.

File: src/ImVkMemorySharedImage.cpp (hunk 6 above)
Comment (ImCRV) [reply src/ImVkMemorySharedImage.cpp:589]: We could remove it or manage it through CMake if we want to keep the possibility to build without GL interop (and DX interop).

File: src/vulkan/ImVkCore.hpp (hunk 1 above)
Comment (robin-imverse) [reply src/vulkan/ImVkCore.hpp:132]: ok good

**Review by robin-imverse — APPROVED** [review]

**Pull request merged**